ceno_zkvm/src/scheme/mock_prover.rs

```diff
@@ -404,58 +456,55 @@ impl<'a, E: ExtensionField + Hash> MockProver<E> {
         wits_in: &[ArcMultilinearExtension<'a, E>],
         challenge: [E; 2],
         lkm: Option<LkMultiplicity>,
-    ) -> Result<(), Vec<MockProverError<E>>> {
+    ) -> Result<(), Vec<MockProverError<E, u64>>> {
         Self::run_maybe_challenge(cb, wits_in, &[], &[], Some(challenge), lkm)
     }
 
     pub fn run(
         cb: &CircuitBuilder<E>,
         wits_in: &[ArcMultilinearExtension<'a, E>],
-        programs: &[ceno_emul::Instruction],
+        program: &[ceno_emul::Instruction],
         lkm: Option<LkMultiplicity>,
-    ) -> Result<(), Vec<MockProverError<E>>> {
-        Self::run_maybe_challenge(cb, wits_in, programs, &[], None, lkm)
+    ) -> Result<(), Vec<MockProverError<E, u64>>> {
+        Self::run_maybe_challenge(cb, wits_in, program, &[], None, lkm)
     }
 
     fn run_maybe_challenge(
         cb: &CircuitBuilder<E>,
         wits_in: &[ArcMultilinearExtension<'a, E>],
-        input_programs: &[ceno_emul::Instruction],
+        program: &[ceno_emul::Instruction],
         pi: &[ArcMultilinearExtension<'a, E>],
         challenge: Option<[E; 2]>,
         lkm: Option<LkMultiplicity>,
-    ) -> Result<(), Vec<MockProverError<E>>> {
-        let program = Program::new(
-            CENO_PLATFORM.pc_base(),
-            CENO_PLATFORM.pc_base(),
-            input_programs.to_vec(),
-            Default::default(),
-        );
+    ) -> Result<(), Vec<MockProverError<E, u64>>> {
+        let program = Arc::new(Program::from_insn_codes(program));
+        let (table, challenge) = Self::load_tables_with_program(cb.cs, program, challenge);
 
-        // load tables
-        let (challenge, mut table) = if let Some(challenge) = challenge {
-            (challenge, load_tables(cb, challenge))
-        } else {
-            load_once_tables(cb)
-        };
-        let mut prog_table = vec![];
-        Self::load_program_table(&mut prog_table, &program, challenge);
-        for prog in prog_table {
-            table.insert(prog);
-        }
+        Self::run_with_challenge_plain(cb.cs, &table, wits_in, pi, 1, challenge, lkm, None)
+    }
 
+    #[allow(clippy::too_many_arguments)]
+    fn run_with_challenge_plain(
+        cs: &ConstraintSystem<E>,
+        table: &HashSet<Vec<u64>>,
+        wits_in: &[ArcMultilinearExtension<'a, E>],
+        pi: &[ArcMultilinearExtension<'a, E>],
+        num_instances: usize,
+        challenge: [E; 2],
+        lkm: Option<LkMultiplicity>,
```

Comment:
not related to this PR, but to differentiate with `shared_lkm`, how about rename `lkm` as `expected_lkm`

Reply:
I've renamed it.